### ISSUE TITLE
error: inlining failed in call to 'salsa20'

### ISSUE THREAD

**Issue by andreas-schwab** [besser82/libxcrypt#95]
libxcrypt build fails on i586 with gcc 10:

alg-yescrypt-opt.c: In function 'blockmix_salsa8':
alg-yescrypt-opt.c:280:20: error: inlining failed in call to 'salsa20': --param max-inline-insns-single limit reached [-Werror=inline]
  280 | static inline void salsa20(salsa20_blk_t *restrict B,
      |                    ^~~~~~~
alg-yescrypt-opt.c:364:2: note: called from here
  364 |  salsa20(&X, &out, 4);
      |  ^~~~~~~~~~~~~~~~~~~~
alg-yescrypt-opt.c:383:3: note: in expansion of macro 'SALSA20_8_XOR_MEM'
  383 |   SALSA20_8_XOR_MEM(Bin[i * 2 + 1], Bout[r + i])
      |   ^~~~~~~~~~~~~~~~~
alg-yescrypt-opt.c:280:20: error: inlining failed in call to 'salsa20': --param max-inline-insns-single limit reached [-Werror=inline]
  280 | static inline void salsa20(salsa20_blk_t *restrict B,
      |                    ^~~~~~~
alg-yescrypt-opt.c:364:2: note: called from here
  364 |  salsa20(&X, &out, 4);
      |  ^~~~~~~~~~~~~~~~~~~~
alg-yescrypt-opt.c:382:3: note: in expansion of macro 'SALSA20_8_XOR_MEM'
  382 |   SALSA20_8_XOR_MEM(Bin[i * 2], Bout[i])
      |   ^~~~~~~~~~~~~~~~~
alg-yescrypt-opt.c: In function 'blockmix_salsa8_xor':
alg-yescrypt-opt.c:280:20: error: inlining failed in call to 'salsa20': --param max-inline-insns-single limit reached [-Werror=inline]
  280 | static inline void salsa20(salsa20_blk_t *restrict B,

**Comment by solardiz** [comment]
Possible fix:

```c
#ifdef __GNUC__
#define inline __attribute__((always_inline)) inline
#endif
```

More elaborate version of this from JtR, but I think this is excessive portability for libxcrypt (inconsistent with the rest of it):

```c
#ifdef __GNUC__
#if __GNUC__ > 3 || (__GNUC__ == 3 && __GNUC_MINOR__ >= 1)
#define MAYBE_INLINE __attribute__((always_inline)) __inline__
#else
#define MAYBE_INLINE __inline__
#endif
#elif __STDC_VERSION__ >= 199901L
#define MAYBE_INLINE inline
#else
#define MAYBE_INLINE
#endif
```

and use `MAYBE_INLINE` in place of `inline` further.

Please also see how yescrypt's `sha256.c` takes care of the `restrict` keyword. It does bother checking `__STDC_VERSION__ >= 199901L` as well as using gcc's `__restrict`. This is probably excessive within libxcrypt, but it does mean I'm likely to use more elaborate logic in upstream yescrypt in both places for consistency, and libxcrypt would ideally not differ from upstream unnecessarily.

**Comment by zackw** [comment]
I'm inclined to fix this by turning off `-Winline`:

```
--- a/m4/zw_simple_warnings.m4
+++ b/m4/zw_simple_warnings.m4
@@ -72,7 +72,6 @@ AC_ARG_ENABLE(
         -Wformat-overflow=2 dnl
         -Wformat-signedness dnl
         -Wformat-truncation=1 dnl
-        -Winline dnl
         -Wlogical-op dnl
         -Wmissing-declarations dnl
         -Wmissing-prototypes dnl
```

@andreas-schwab Can you confirm that that fixes the build with GCC 10?

@solardiz Is there any reason why the functions in alg-yescrypt-opt.c really _must_ be inlined?  A correctness reason, not a performance reason?

**Comment by solardiz** [comment]
@zackw There's no correctness reason, only performance and lack of significant code size increase. When gcc refuses to inline these, I guess it might be considering the inlined function's size only and not considering the savings from not needing to pass parameters and the specialization of the function that inlining ultimately enables.

**Comment by solardiz** [comment]
Regardless, I don't mind libxcrypt dropping `-Winline`, and I'll look into fixing this upstream separately.

**Comment by zackw** [comment]
Thinking about it a bit more, even if you make changes to yescrypt to avoid this warning, we should drop `-Winline`.  We use aggressive warning settings and `-Werror` in all builds, because we want to fail the build if some unusual environment exposes a latent bug.  But `-Winline` isn't like that; it diagnoses compiler limitations and potential efficiency problems, not potential bugs.  We're just causing trouble for repackagers by having it in there.

Please do send a PR for whatever changes you end up making to yescrypt upstream.

**Comment by zackw** [comment]
@andreas-schwab I tested with GCC 10 myself.  You will need both 45ff874 and the subsequent 941362e for a successful build.

**Comment by besser82** [comment]
Another issue with GCC 10.x on s390x only has been fixed in 0eb14f1e178a1ba9c897bb62db1f5ced3e6d9ab8.
